building_sim_plugins/building_plugins_common/src/slotcar_common.cpp

```diff
@@ -454,3 +558,87 @@ void SlotcarCommon::map_cb(
   _initialized_levels = true;
 
 }
+
+// Enables/disables charge when called from Ignition/Gazebo plugin
+void SlotcarCommon::charge_state_cb(
+  const std::string& name, bool selected)
+{
+  if (name == _enable_charge_str)
+  {
+    _enable_charge = selected;
+  }
+  else if (name == _enable_instant_charge_str)
+  {
+    _enable_instant_charge = selected;
+  }
+  else if (name == _enable_drain_str)
+  {
+    _enable_drain = selected;
+  }
+  else
+  {
+    std::cerr << "Invalid button selected. " << std::endl;
+  }
+}
+
+bool SlotcarCommon::near_charger(const Eigen::Isometry3d& pose) const
+{
+  std::string lvl_name = get_level_name(pose.translation()[2]);
+  auto waypoints_it = _charger_waypoints.find(lvl_name);
+  if (waypoints_it != _charger_waypoints.end())
+  {
+    const std::vector<ChargerWaypoint>& waypoints = waypoints_it->second;
+    for (const ChargerWaypoint& waypoint : waypoints)
+    {
+      // Assumes it is on the same Z-plane
+      double dist =
+        sqrt(pow(waypoint.x - pose.translation()[0], 2)
+          + pow(waypoint.y - pose.translation()[1], 2));
+      if (dist < _charger_dist_thres)
+      {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+double SlotcarCommon::compute_charge(const double run_time) const
+{
+  const double dQ = _params.charging_current * run_time; // Coulombs
+  const double dSOC = dQ / (_params.nominal_capacity * 3600.0);
+  return dSOC;
+}
+
+double SlotcarCommon::compute_discharge(
+  const Eigen::Vector3d lin_vel, const double ang_vel,
+  const Eigen::Vector3d lin_acc, const double ang_acc,
+  const double run_time) const
+{
+  const double v = lin_vel.norm();
+  const double w = std::abs(ang_vel);
+  const double a = lin_acc.norm();
+  const double alpha = std::abs(ang_acc);
+
+  // Loss through acceleration
+  const double EA = ((_params.mass * a * v) +
+    (_params.inertia * alpha * w)) * run_time;
+
+  // Loss through friction. Ignores friction due to rotation in place
+  const double EF = compute_friction_energy(_params.friction_coefficient,
+      _params.mass, v, run_time);
+
+  // Change in energy as a result of motion
+  const double dE = EA + EF;
```

Comment:
Can we tidy up this part by renaming `dE2` to dED` for energy drain from devices and adding the same to `dE` in this step.